File: src/com/medium/LongestPalindromicSubstring_5.java
/**
 * 
 */
package com.medium;


/**
 * @author Chris Lin
 * @version 1.0
 */
public class LongestPalindromicSubstring_5 {
    // 2019年8月6日
    // Use valid palindrome method
    // Time Complexity: O(n^2)
    // Space Complexity:O(1)
    // Runtime: 5 ms, faster than 96.59%
    // Memory Usage: 36.2 MB, less than 100.00%
    private int mMaxLength = 0;
    private int mIndexStart = 0;

    public String longestPalindrome(String s) {
        if (s.length() < 2) {
            return s;
        }
        int mLength = s.length();

        for (int i = 0; i < mLength - 1; i++) {
            searchPalindrome(s, i, i);
            searchPalindrome(s, i, i + 1);
        }
        return s.substring(mIndexStart, mIndexStart + mMaxLength);
    }

    private void searchPalindrome(String s, int mIndexLeft, int mIndexRight) {
        while (mIndexLeft >= 0 && mIndexRight < s.length()) {
            if (s.charAt(mIndexLeft) != s.charAt(mIndexRight)) {
                break;
            }
            mIndexLeft--;
            mIndexRight++;
        }
        if (mMaxLength < mIndexRight - mIndexLeft - 1) {
            mMaxLength = mIndexRight - mIndexLeft - 1;
            mIndexStart = mIndexLeft + 1;
        }
    }

    // 2019年8月6日
    // Use DP
    // Time Complexity: O(n^2)
    // Space Complexity:O(n^2)
    // Runtime: 32 ms, faster than 45.97%
    // Memory Usage: 38.1 MB, less than 43.41%
    public String longestPalindrome2(String s) {
        // if s[i] == s[j] , mDP[i][j] = true
        if (s == null || s.length() == 0) {
            return s;
        }
        int mLength = s.length();
        int mIndexStart = 0, mMaxLength = 0;
        // mDP[i][j] indicates whether substring s starting at index i and
        // ending at j is palindrome
        boolean[][] mDP = new boolean[mLength][mLength];

        for (int i = mLength - 1; i >= 0; i--) {
            for (int j = i; j < mLength; j++) {
                mDP[i][j] = s.charAt(i) == s.charAt(j) && (j - i < 3 || mDP[i + 1][j - 1]);

                if (mDP[i][j] && (j - i + 1 > mMaxLength)) {
                    mIndexStart = i;
                    mMaxLength = j - i + 1;
                }
            }
        }
        return s.substring(mIndexStart, mIndexStart + mMaxLength);
    }
}

File: Golang_src/medium/longestPalindromicSubstring_5.go
package medium

// Use brute force:
// Time Complexity: O(n^3)
// Space Complexity:O(1)
// Runtime: 164 ms, faster than 31.64%
// Memory Usage: 2.7 MB, less than 48.91%
func longestPalindrome(s string) string {
	size := len(s)
	result := ""
	for i := 0; i < size; i++ {
		for j := i; j < size; j++ {
			if j-i+1 > len(result) && isPalindrome(s[i:j+1]) {
				result = s[i : j+1]
			}
		}
	}
	return result
}

func isPalindrome(s string) bool {
	size := len(s)
	for i := 0; i < size/2+1; i++ {
		if s[i] != s[size-i-1] {
			return false
		}
	}
	return true
}

// Use center extend:
// Time Complexity: O(n^2)
// Space Complexity:O(1)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 0.0 MB, less than 100%
func longestPalindrome2(s string) string {
	size := len(s)
	length := 0
	start := 0

	getLen := func(s string, l, r int) int {
		for l >= 0 && r < len(s) && s[l] == s[r] {
			l--
			r++
		}
		// due to out of range of l & r,
		// currently we need (l+1) ~ (r-1)
		return (r - 1) - (l + 1) + 1
	}

	for i := 0; i < size; i++ {
		cur := max(getLen(s, i, i), getLen(s, i, i+1))
		if cur > length {
			length = cur
			start = i - (cur-1)/2
		}
	}

	return s[start : start+length]
}

// Use center extend 2:
// Time Complexity: O(n^2)
// Space Complexity:O(1)
// Runtime: 13 ms, faster than 58.97%
// Memory Usage: 2.9 MB, less than 47.92%
func longestPalindrome3(s string) string {

	var dfs func(l, r int) (int, int)
	dfs = func(l, r int) (int, int) {
		for l >= 0 && r < len(s) {
			if s[l] != s[r] {
				break
			}
			l--
			r++
		}

		return l + 1, r - 1
	}

	left, right := 0, 0

	for i := 1; i < len(s); i++ {
		l1, r1 := dfs(i, i)
		if r1-l1 > right-left {
			left, right = l1, r1
		}
		l2, r2 := dfs(i-1, i)
		if r2-l2 > right-left {
			left, right = l2, r2
		}
	}

	return s[left : right+1]
}
